File: LC 215.cpp
class Solution {
public:
    int findKthLargest(vector<int>& nums, int k) {
        int n = nums.size();
        function<int(int, int)> quickSort = [&] (int i, int j) -> int {
            int pos = i, pvt = nums[i];
            int l = i, r = j;
            while (l < r) {
                while (l < r && nums[r] > pvt) {
                    r--;
                }
                while (l < r && nums[l] <= pvt) {
                    l++;
                } 
                swap(nums[l], nums[r]);
            }
            swap(nums[pos], nums[l]);
            if (l == n - k) {
                return nums[l];
            } else if (l < n - k) {
                return quickSort(l+1, j);
            } else {
                return quickSort(i, l-1);
            }
        };
        return quickSort(0, n-1);
    }
};

File: 1000/LC 215.go
func findKthLargest(nums []int, k int) int {
    n := len(nums)
    rand.Seed(time.Now().UnixNano())

    var quickSort func(int, int) int
    quickSort = func(l, r int) int {
        if l >= r {
            return nums[l]
        }
        pivot := nums[l + rand.Intn(r - l + 1)]
        i, j, m := l, l, r + 1
        for i < m {
            if nums[i] > pivot {
                m--
                nums[i], nums[m] = nums[m], nums[i]
            } else if nums[i] < pivot {
                nums[i], nums[j] = nums[j], nums[i]
                i++
                j++
            } else {
                i++
            }
        }

        if n - j < k {
			return quickSort(l, j-1)
		} else if n - m >= k {
			return quickSort(m, r)
		}
		return nums[j]
    }

    return quickSort(0, n - 1)
}
